### PR TITLE
refactor: wraper RocksdbIterator to refactor PartitionIterator

### PR DESCRIPTION
**Pull request by foreverneverer**

### CODE REVIEW

File: pegasus-spark-analyser/src/main/java/com/xiaomi/infra/pegasus/spark/analyser/ColdBackupLoader.java [hunk 1]
```diff
@@ -49,24 +52,28 @@ public ColdBackupLoader(ColdBackupConfig config) throws PegasusSparkException {
...
-
   @Override
   public int getPartitionCount() {
     return partitionCount;
   }
 
   @Override
-  public Map<Integer, String> getCheckpointUrls() {
-    return checkpointUrls;
+  public ColdBackupConfig getConfig() {
```
Comment (levy5307) [inline pegasus-spark-analyser/src/main/java/com/xiaomi/infra/pegasus/spark/analyser/ColdBackupLoader.java:61]: `public final ColdBackupConfig getConfig() {`

Comment (foreverneverer) [reply pegasus-spark-analyser/src/main/java/com/xiaomi/infra/pegasus/spark/analyser/ColdBackupLoader.java:61]: `final` method means the method don't be overload(different from c++ const), the method is abstract in interface, it should't be `final`

File: pegasus-spark-analyser/src/main/java/com/xiaomi/infra/pegasus/spark/analyser/ColdBackupLoader.java [hunk 2]
```diff
@@ -49,24 +52,28 @@ public ColdBackupLoader(ColdBackupConfig config) throws PegasusSparkException {
...
+    return coldBackupConfig;
   }
 
   @Override
-  public PegasusRecord restoreRecord(RocksIterator rocksIterator) {
-    return coldBackupConfig.getDataVersion().getPegasusRecord(rocksIterator);
+  public PegasusScanner getScanner(int pid) throws PegasusSparkException, RocksDBException {
```
Comment (neverchanje) [inline pegasus-spark-analyser/src/main/java/com/xiaomi/infra/pegasus/spark/analyser/ColdBackupLoader.java:66]:
```suggestion
  public PegasusScanner getColdBackupScanner(int pid) throws PegasusSparkException, RocksDBException {
```

File: pegasus-spark-analyser/src/main/java/com/xiaomi/infra/pegasus/spark/analyser/PegasusScanner.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package com.xiaomi.infra.pegasus.spark.analyser;
+
+public interface PegasusScanner {
```
Comment (neverchanje) [inline pegasus-spark-analyser/src/main/java/com/xiaomi/infra/pegasus/spark/analyser/PegasusScanner.java:3]: It shouldn't be a public interface. The user must always use the RDD abstraction provided by pegasus-analyzer. `PegasusScanner` is hidden under RDD.

File: pegasus-spark-analyser/src/main/java/com/xiaomi/infra/pegasus/spark/analyser/ColdBackupLoader.java [hunk 4]
```diff
@@ -160,4 +167,49 @@ private int getCount(String prefix) throws PegasusSparkException {
     throw new PegasusSparkException(
         "can't find the partition count failed, [url: " + appMetaDataUrl + "]");
   }
+
+  public static class ColdBackupScanner implements PegasusScanner {
```
Comment (neverchanje) [inline pegasus-spark-analyser/src/main/java/com/xiaomi/infra/pegasus/spark/analyser/ColdBackupLoader.java]:
```suggestion
  static class ColdBackupScanner implements PegasusScanner {
```

File: pegasus-spark-analyser/src/main/java/com/xiaomi/infra/pegasus/spark/analyser/PegasusScanner.java (hunk 3 above)
Comment (foreverneverer) [reply pegasus-spark-analyser/src/main/java/com/xiaomi/infra/pegasus/spark/analyser/PegasusScanner.java:3]: java `Interface` default public and need public which meas can be inherited and overload. Of course, you can put one interface into another class to hide it. But I think it's not necessary.

File: pegasus-spark-analyser/src/main/java/com/xiaomi/infra/pegasus/spark/analyser/ColdBackupLoader.java (hunk 2 above)
Comment (foreverneverer) [reply pegasus-spark-analyser/src/main/java/com/xiaomi/infra/pegasus/spark/analyser/ColdBackupLoader.java:66]: `getScanner` is abstract method in `PegasusScanner` which need overload to offer different `Scanner`, For example, we can add `OnlineReader` and offer `OnlineScanner`

**Review by neverchanje — APPROVED** [review]

**Pull request merged**